### DISCUSSION TITLE
A Query Logger

### DISCUSSION THREAD

**Discussion by Mamaduka** [spatie/wordpress-ray#1] [Q&A]
Hi,

I'm planning to start working on WP specific query logger implementation. While I've technical details for the WP part, I'm not sure what's the best practice for adding this feature to this plugin.

The class itself will be similar to Laravel Ray's `QueryLogger`.

The documentation only shows simple extending examples, using marcoables.  But in this case, I think a class base approach would be preferable. Mostly to share logger instance between `showQueries()` and `stopShowingQueries()` methods.

Any examples would be helpful.

Thanks,

George

**Comment by freekmurze** [comment]
Hi, 

thanks for wanting to help out. Macro's are not the way to go indeed.

I think we should create a `Payloads` directory and copy over the ExecutedQueryPayload from the laravel-ray package: https://github.com/spatie/laravel-ray/blob/c69db7ec062c0c98af8caac207eae4fde9ef6b5d/src/Payloads/ExecutedQueryPayload.php

Next, a query logger similar to [this one](https://github.com/spatie/laravel-ray/blob/3b6f7769d6fab0644c47c2544aea963c6c530c25/src/QueryLogger.php#L9) from the Laravel implementation should be added.

If have any more questions, or need more help let me know.

This will be helpful to a lot of WordPress devs, and I greatly appreciate your work on this!

**Reply by Mamaduka** [reply]
Hello,

> I think we should create a Payloads directory and copy over the ExecutedQueryPayload from the laravel-ray package: https://github.com/spatie/laravel-ray/blob/c69db7ec062c0c98af8caac207eae4fde9ef6b5d/src/Payloads/ExecutedQueryPayload.php
> 
> Next, a query logger similar to this one from the Laravel implementation should be added.

This is my plan as well and I already have a prototype.

The only thing I'm not sure about is how to do this non-Marco way. Or how to structure the main `Spatie\WordpressRay\Ray` class. If you have time to provide a base/skeleton, I can build on top of it.

Prototype screenshot:
<img width="700" alt="wp-query-logger" src="https://user-images.githubusercontent.com/240569/104158166-3aebc900-5406-11eb-8c76-89c03df0fc3f.png">

**Reply by freekmurze** [reply]
Could you push the code that you already have? To make things easy, I'll give you write access to the repo, so you can commit on the master branch. I'll polish the code you already have.

We should also make the origin work. Right now the origin displayed in `class-wp-hook`, but this should be the actual place where the query was executed. In Laravel we do this by registering an custom `OriginFactory` and do any overrides there. https://github.com/spatie/laravel-ray/blob/0c8d7c825b59103121227c591c43276062b1f8ef/src/OriginFactory.php#L3

**Reply by Mamaduka** [reply]
Done - 829bdab.

While prototyping, I was using the `QueryLogger` class directly (not ideal).

I will look into `OriginFactory` later today.

**Reply by freekmurze** [reply]
Thank you so much, I'll review it all this evening!

**Reply by freekmurze** [reply]
Thanks for the code, I've rewritten it for clarity.

You can now enable the functionality with `ray()->showQueries()` 

I haven't tested it yet. Feel free to PR improvements!

**Reply by Mamaduka** [reply]
I think we can close this one out.

P.S. I pushed few minor fixes to the logger, hope you don't mind.

**Reply by freekmurze** [reply]
Thanks for the fixes.

Discussions can't be closed, but I also think the functionality is ok now 👍
Thanks for your help with this!
